ntp-proto/src/filter.rs

```diff
@@ -0,0 +1,283 @@
+        let jitter = root_mean_square / (register.len() - 1) as f64;
+
+        // In order to ensure consistency and avoid divide exceptions in other
+        // computations, the psi is bounded from below by the system precision
+        // s.rho expressed in seconds.
+        jitter.max(s.precision)
+    }
+}
+
+pub struct System {
+    precision: f64,
+    poll: NtpDuration,
+    leap_indicator: NtpLeapIndicator,
+}
+
+impl System {
+    #[cfg(test)]
+    fn dummy() -> Self {
+        Self {
+            precision: 0.0,
+            poll: NtpDuration::default(),
+            leap_indicator: NtpLeapIndicator::NoWarning,
+        }
+    }
+}
+
+pub struct Peer {
+    clock_filter: ClockFilterContents,
+    t: NtpTimestamp,
+
+    jitter: f64,
+    dispersion: NtpDuration,
+
+    offset: NtpDuration,
+    delay: NtpDuration,
+
+    burst_counter: u32,
+}
+
+pub struct LocalClock {
+    t: NtpTimestamp,
+}
+
+#[allow(dead_code)]
+pub fn clock_filter(
+    peer: &mut Peer,
+    s: &System,
+    c: &LocalClock,
+    clock_offset: NtpDuration,
+    roundtrip_delay: NtpDuration,
+    dispersion: NtpDuration,
+) {
+    let new_tuple = FilterTuple {
+        offset: clock_offset,
+        delay: roundtrip_delay,
+        dispersion,
+        time: c.t,
+    };
+
+    let dispersion_correction = (c.t - peer.t) / ONE_OVER_PHI;
+    peer.clock_filter
+        .shift_and_insert(new_tuple, dispersion_correction);
+
+    let mut temporary_list = peer.clock_filter;
+
+    temporary_list.sort_by_delay();
+
+    let smallest_delay = temporary_list.register[0];
+
+    let dtemp = peer.offset;
+    peer.offset = smallest_delay.offset;
+    peer.delay = smallest_delay.delay;
+
+    // TODO (not in the skeleton as far as I can see)
+    // If the first tuple epoch t_0 is not
+    // later than the last valid sample epoch tp, the routine exits without
+    // affecting the current peer variables.
+
+    peer.dispersion = temporary_list.dispersion();
+    peer.jitter = temporary_list.jitter(s, smallest_delay);
+
+    // Prime directive: use a sample only once and never a sample
+    // older than the latest one, but anything goes before first
+    // synchronized.
+    if smallest_delay.time - peer.t <= NtpDuration::default() && s.leap_indicator.is_synchronized()
+    {
+        return;
+    }
+
+    // Popcorn spike suppressor.  Compare the difference between the
+    // last and current offsets to the current jitter.  If greater
+    // than SGATE (3) and if the interval since the last offset is
+    // less than twice the system poll interval, dump the spike.
+    // Otherwise, and if not in a burst, shake out the truechimers.
+    let too_soon = (smallest_delay.time - peer.t) < (s.poll * 2i64);
+    if (peer.offset - dtemp).to_seconds().abs() > SGATE * peer.jitter && too_soon {
+        return;
+    }
+
+    peer.t = smallest_delay.time;
+    if peer.burst_counter == 0 {
+        todo!()
+        // clock_select();
```

Comment:
yes, hence the `todo!` because it turns out that we want to do this very differently from what the reference is doing, for the reasons you describe. 

How to do it was harder to see at the time, but now that I've looked at clock selection in more detail we indeed should make this have its own (smaller) state, passing in all of `peer` and `s` is just confusing